### PR TITLE
Better caching

### PR DESCRIPTION
**Pull request by borisyankov**
This is based on previous caching PR. Ready to merge after review.

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@borisyankov, it looks like you've already signed the Dropbox CLA. Thanks!

File: src/chat/__tests__/chatReducers-test.js [hunk 1]
```diff
@@ -0,0 +1,345 @@
+import messagesReducers from '../chatReducers';
```
Comment (neerajwahi) [inline src/chat/__tests__/chatReducers-test.js]: Maybe rename this to `chatReducers` too?

File: src/message-list/MessageList.js [hunk 2]
```diff
@@ -29,10 +29,9 @@ export default class MessageList extends React.PureComponent {
         style={styles.list}
         contentContainerStyle={styles.container}
         automaticallyAdjustContentInset="false"
-        autoScrollToBottom={caughtUp}
+        autoScrollToBottom
```
Comment (neerajwahi) [inline src/message-list/MessageList.js]: Please leave `caughtUp` in the codebase. I need to fix its behavior but we'll need it

File: src/message-list/messagesActions.js [hunk 3]
```diff
@@ -1,36 +1,33 @@
...
 
     dispatch({
       type: MESSAGE_FETCH_SUCCESS,
       auth,
       messages,
       anchor,
       narrow,
-      caughtUp: (numAfter >= numBefore &&
-        messages.length < numAfter + numBefore + (numBefore ? 1 : 0)),
+      startReached: numAfter === 0 && numBefore > messages.length,
```
Comment (neerajwahi) [inline src/message-list/messagesActions.js]: Please leave `caughtUp`

File: src/chat/chatReducers.js [hunk 4]
```diff
@@ -1,29 +1,100 @@
...
+      return {
+        ...state,
+        narrow: action.narrow
+      };
     case MESSAGE_FETCH_START:
       return {
         ...state,
-        fetching: true,
-        narrow: action.narrow,
+        fetching: state.fetching + 1,
```
Comment (neerajwahi) [inline src/chat/chatReducers.js]: Why is this an integer instead of a boolean now?

Comment (borisyankov) [inline src/chat/chatReducers.js]: Because we might have two fetching at the same time.

File: src/message-list/messagesActions.js (hunk 3 above)
Comment (borisyankov) [inline src/message-list/messagesActions.js]: Will need to rework it in the way 'startReached' is - per cached narrow. Not much harder, but different.

**Pull request merged**